File: csclub/2013/apr19/maxdist/maxdist.py
import sys
from itertools import izip
n,d=map(int,raw_input().split())
m=[map(int,sys.stdin.readlines())]
M=m[:]
while m[-1]:
	m.append(map(min,izip(m[-1],m[-1][2**(len(m)-1):])))
	M.append(map(max,izip(M[-1],M[-1][2**(len(M)-1):])))
best,start,end,lb=0,0,1,0
while end<=n:
	#lb=int(log(end-start,2)) is preferred if the language supports it
	exp=2**lb#highest set bit=2**(binary logarithm)
	if max(M[lb][start],M[lb][end-exp])-min(m[lb][start],m[lb][end-exp])<=d:
		best=max(best,end-start)
		end+=1
		if end-start>=2*exp:
			lb+=1
	else:
		start+=1
		if end-start<exp:
			lb-=1
print best

File: csclub/2013/apr19/maxdist/maxdist.cpp
#include<stdio.h>
#include<algorithm>
using namespace std;
int main(){
	int n,d;
	scanf("%d%d",&n,&d);
	int a[n+1];
	for(int i=0;i<n;scanf("%d",&a[i++]));
	a[n]=a[n-1];
	int best=0;
	for(int start=0;start<n;++start){
		int m=a[start],M=a[start],end=start;
		for(;M-m<=d&&end<=n;++end){
			m=min(m,a[end]);
			M=max(M,a[end]);
		}
		best=max(best,end-1-start);
	}
	printf("%d\n",best);
	return 0;
}
